Started draft for user guide (#21)
nephio-project/docs: user-guide/user-guide.md

Nephio breaks down the larger problem into two primary areas:

* Kubernetes as a uniform automation control plane in each site to configure all aspects of the distributed cloud and network functions
* An automation framework that leverages Kubernetes declarative, actively-reconciled methodology along with machine-manipulable configuration to tame the complexity of these configurations.

### Kubernetes as a Uniform Automation Control Plane
Utilizing Kubernetes as the automation control plane at each layer of the stack simplifies the overall automation and enables declarative management with active reconciliation for the entire stack. We can broadly think of three layers in the stack, as shown
* Cloud infrastructure
* Workload (network function) resources
* Workload (network function) configuration.
Nephio is establishing open, extensible Kubernetes Custom Resource Definition (CRD) models for each layer of the stack, in conformance to the 3GPP & O-RAN standards.

![nephio-overview.png](nephio-overview.png)

##  Overview of underlying technologies
### Custom Resources and Controllers

> A Custom Resource Definition or CRD is a Kubernetes extension mechanism for adding custom data types to Kubernetes. The CRDs are the schemas - analogous to table definitions in a relational database for example. The instances of those - analogous to rows in a RDBMS - are called Custom Resources or CRs. People often accidentally say "CRDs" when they mean "CRs", so be sure to ask for clarification if the context doesn't make it clear which is meant.

Reviewer: ```suggestion
A Custom Resource Definition (CRD) is an extension mechanism for adding custom data types to Kubernetes. The CRDs are the schemas - analogous to table definitions in a relational database for example. The instances of those - analogous to rows in a RDBMS - are called Custom Resources (CRs). People often accidentally say "CRDs" when they mean "CRs", so be sure to ask for clarification if the context doesn't make it clear which is meant.
```

All Kubernetes resources have a metadata field that contains the name, namespace (if not cluster scoped), annotations and labels for the resource. Most resources also have a spec and a status field. The spec field holds the intended state, and the status field holds the observed state and any control data needed by the controller to do its job. Typically, controllers read the spec and write the status.
### KPT

> KPT is a GitOps based package management tool for kubernetes. It uses configuration as data principle. For more details refer to [KPT book](https://kpt.dev/book/)

Reviewer: ```suggestion
KPT is a GitOps based package management tool for Kubernetes. It uses configuration-as-data principle. For more details refer to [KPT book](https://kpt.dev/book/)
```

### Porch  

> [Package Orchestration service](https://kpt.dev/book/08-package-orchestration/) (Porch) enables workflows similar to those supported by the kpt CLI, but makes them available as a kubernetes service. For more details refer to [Porch user guide](https://kpt.dev/guides/porch-user-guide)

Reviewer: ```suggestion
[Package Orchestration service](https://kpt.dev/book/08-package-orchestration/) (Porch) enables workflows similar to those supported by the kpt CLI, but makes them available as a Kubernetes service. For more details refer to [Porch user guide](https://kpt.dev/guides/porch-user-guide)
```

### ConfigSync 
Config Sync lets cluster operators and platform administrators deploy consistent configurations and policies. You can deploy these configurations and policies to individual Kubernetes clusters, multiple clusters that can span hybrid and multi-cloud environments, and multiple namespaces within clusters. This process simplifies and automates configuration and policy management at scale. Config Sync also lets development teams independently manage their namespaces within clusters, while still being subject to policy guardrails set by administrators.

Config Sync is an open-source project and the source and releases available [here](https://www.github.com/GoogleContainerTools/kpt-config-sync). 

### Packages 

> Packages or Kpt Packages are bundles of Kubernetes resource files, along with a Kptfile (also in Kubernetes Resource Model or KRM format). They provide the basic unit of management in the Kpt toolchain. This toolchain is used to manage the configuration before it reaches the Kubernetes API server. This "shift left" model is critical to allowing safe collaborative, automated configuration creation and editing, because errors or partial configurations can be resolved prior to affecting operations.

Reviewer: ```suggestion
Packages or Kpt Packages are bundles of Kubernetes resource files, along with a Kptfile (also in Kubernetes Resource Model (KRM) format). They provide the basic unit of management in the Kpt toolchain. This toolchain is used to manage the configuration before it reaches the Kubernetes API server. This "shift left" model is critical to allowing safe collaborative, automated configuration creation and editing, because errors or partial configurations can be resolved prior to affecting operations.
```

Stepping up the power and complexity level, the fact that Porch is a Kubernetes API service means that we can build Kubernetes controllers on top of it. This is exactly what we are doing in Nephio.

### Package Ancestry
A package may have a single upstream parent, and many downstream descendants. The Kptfiles in these packages are used to maintain the relationships, capturing ancestry relationships like those shown below.

![package-ancestry.png](package-ancestry.png)

By tracking these relationships, changes at the original source can be propagated via controlled automation down the tree.

### Package Configuration Journey

![package-lifecycle.png](package-lifecycle.png)

## R1 scope
### API
CRDs provided for UPF, SMF and AMF 5g core services
### Web UI
Basic web UI to view and manage Package Variants and Package variant sets.
### Packages
* Kpt packages for all [free5gc](https://free5gc.org/) services
* Packages for core Nephio services
* Packages for cluster API services for cluster creation
* Packages for dependent services
### Functionalities

> * Create kubernetes clusters. This functionality ia based on cluster API. At this time only KIND clusters creation is supported.

Reviewer: ```suggestion
* Create Kubernetes clusters. This functionality is based on [cluster API](https://cluster-api.sigs.k8s.io/). At this time only [KIND](https://kind.sigs.k8s.io/) clusters creation is supported.
```

* Fully automated deployment of UPF, SMF and AMF services of [free5Gc](https://free5gc.org/) . These are deployed on multiple clusters based on user's intent expressed via CRDs. 
* Inter cluster networking setup.
* Deployment of other free5gc functions. Some manual configuration such as IP addresses may be needed for these services. 
* Auto-scale up of UPF, SMF and AMF services based on changes to capacity requirements expressed as user intent. 
 

## Deployment model 
For the purposes of Nephio deployment we can categorize clusters into two varieties

> * Management Cluster
> * Workload cluster
> 
> Management cluster is where the majority of Nephio components are installed. The management cluster is dedicated to manage the deployment and life cycle management of network functions that will be deployed on workload clusters.
> 
> Workload cluster is where the actual network function workloads are deployed and running.

Reviewer: ```suggestion
* Management Cluster: Where the majority of Nephio components are installed. It is dedicated to manage the deployment and life cycle management of network functions that will be deployed on workload clusters.
* Workload Cluster: Where the actual network function workloads are deployed and running.
```

The diagram below depicts deployment at the high level.

![nephio-deployment.png](nephio-deployment.png)

### Management cluster components

#### Infrastructure Components
* Porch
* ConfigSync

#### Nephio controllers
* Nephio Controller Operator
* Vlan and IPAM Controller
* Cluster API Infrastructure controller
* Status Aggregator  

#### Specializers and Injectors (KRM Functions)
* VLAN 
* IPAM

> * NAD
> * DNN 

Reviewer: ```suggestion
* NAD - NetworkAttachmentDefinition
* DNN - DataNetwork
```
